### PR TITLE
test/e2e: add framework for e2e tests

### PR DESCRIPTION
**Pull request by AlexNPavel**
This will allow us to add more tests in the future without
duplicating code and will also allow us to parse command line options
more easily.

This is one of the parts of PR #355 that is being split off for easier review.

### CODE REVIEW

File: test/e2e/framework/main_entry.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	"testing"
+)
+
+func MainEntry(m *testing.M) {
+	if err := setup(); err != nil {
+		log.Fatalf("Failed to set up framework: %v", err)
+	}
+
+	code := m.Run()
+	os.Exit(code)
```
Comment (fanminshi) [inline test/e2e/framework/main_entry.go]: you can do `os.Exit(m.Run())` to save a line.

File: test/e2e/framework/framework.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
+package framework
+
+import (
+	"flag"
+	"os"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+
```
Comment (fanminshi) [inline test/e2e/framework/framework.go]: no need to have extra line here.

File: test/e2e/framework/framework.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+var Global *Framework
+
+type Framework struct {
+	KubeConfig *rest.Config
+	KubeClient kubernetes.Interface
+}
+
+func setup() error {
+	config := flag.String("kubeconfig", os.Getenv("HOME")+"/.kube/config", "kube config path, e.g. $HOME/.kube/config")
```
Comment (fanminshi) [inline test/e2e/framework/framework.go]: What if `os.Getenv("HOME")` doesn't exist?

Prefer to check env var first `os.LookupEnv("HOME")` before using it in `flag.String(...)`.

**Comment by AlexNPavel** [comment]
Fixed

File: test/e2e/memcached_test.go [hunk 4]
```diff
@@ -7,16 +7,14 @@ import (
 	"net/http"
 	"os"
 	"os/exec"
-	"path/filepath"
 	"testing"
 
 	"github.com/operator-framework/operator-sdk/test/e2e/e2eutil"
+	framework "github.com/operator-framework/operator-sdk/test/e2e/framework"
```
Comment (fanminshi) [inline test/e2e/memcached_test.go]: new line between 13 and 14.

We usually have three groups of imports. The first is the golang internal ones, the second is the imports from this repo, the third is everything else.

File: test/e2e/memcached_test.go [hunk 5]
```diff
@@ -88,21 +89,10 @@ func TestMemcached(t *testing.T) {
...
-	kubeconfig, err := clientcmd.BuildConfigFromFlags("", kubeconfigPath)
-	if err != nil {
-		t.Fatal(err)
-	}
-	kubeclient, err := kubernetes.NewForConfig(kubeconfig)
-	if err != nil {
-		t.Fatal(err)
-	}
 
 	// create namespace
```
Comment (fanminshi) [inline test/e2e/memcached_test.go]: make the variables close to its usage.
e.g:

```go
f := framework.Global
namespace := "memcached" 
// create namespace ...
```

**Comment by AlexNPavel** [comment]
Rebased and fixed

File: test/e2e/framework/framework.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+type Framework struct {
+	KubeConfig *rest.Config
+	KubeClient kubernetes.Interface
+}
+
+func setup() error {
+	homedir, ok := os.LookupEnv("HOME")
+	var config *string
+	if !ok {
```
Comment (fanminshi) [inline test/e2e/framework/framework.go]: Maybe change to following?

```go
defaultKubeConfig := ""
homedir, ok := os.LookupEnv("HOME")
if ok {
 defaultKubeConfig = homedir+"/.kube/config"
}
config := flag.String("kubeconfig", defaultKubeConfig , "kube config path, e.g. $HOME/.kube/config")
```

**Comment by AlexNPavel** [comment]
Fixed

**Comment by fanminshi** [comment]
lgtm defer @hasbro17

File: test/e2e/framework/framework.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	KubeClient kubernetes.Interface
+}
+
+func setup() error {
+	defaultKubeConfig := ""
+	homedir, ok := os.LookupEnv("HOME")
+	if ok {
+		defaultKubeConfig = homedir + "/.kube/config"
+	}
+	config := flag.String("kubeconfig", defaultKubeConfig, "kube config path, e.g. $HOME/.kube/config")
```
Comment (hasbro17) [inline test/e2e/framework/framework.go]: `kube config path, e.g. $HOME/.kube/config` ==> `kubeconfig path, defaults to $HOME/.kube/config`

**Comment by hasbro17** [comment]
LGTM

**Pull request merged**